### ISSUE TITLE
Doubts about the total number of training steps and whether you can open your checkpoint.

### ISSUE THREAD

**Issue by prsigma** [bowang-lab/MorphoDiff#4]
In your training script, the maximum training steps are set to 500. However, in my attempt to reproduce the results on the BBBC021 dataset, the model trained for 500 steps did not learn any morphological changes at all. The resulting images look like this:
![Image](https://github.com/user-attachments/assets/1ddd5f66-468a-4ebc-be77-c859ee4dbfdc)

I would like to ask if your training was also done with only 500 steps? Also, could you please share the checkpoints from your training so that we can reproduce the results in your paper? Thank you!

**Comment by znavidi** [comment]
Hi @prsigma,

We trained our models well beyond 500 steps, with detailed training steps for each experiment provided in Appendix Table 4 of the manuscript (e.g., ~130K steps for RxRx1). The max_train_steps=500 setting in `train.sh` defines the training steps and checkpointing_steps defines when to save checkpoints for one run. These arguments can be adjusted based on your available computing resources and running time.

For training longer, we added the `scontrol requeue $SLURM_JOB_ID` line at the end of the  `train.sh` file to requeue the Slurm job once one run is completed, and resume from the latest checkpoint saved in the`CKPT_LOG_FILE` path. You can uncomment this line if you want the script to resume training for longer, and can monitor the successfully trained checkpoints (in order) in the `CKPT_LOG_FILE` file. I would suggest making sure the train.sh successfully trains for one run and saves checkpoints properly before uncommenting the above line to resume training for longer. 

The sample image you shared is not unreasonable for 500 steps. As training progresses, the model learns cell sparsity and structural patterns more effectively.

You can download the checkpoints from https://huggingface.co/navidi/MorphoDiff_checkpoints/tree/main (also linked in the README). Set the downloaded checkpoint path as `CKPT_PATH` in `generate_img.sh`, adjusting other parameters as needed for your experiment. We've tested these checkpoints, and they work as expected, but please let us know if you encounter any issues—we’d be happy to help!

**Comment by prsigma** [comment]
Thank you for your patient response. In your manuscript, you mentioned that the images used for training on BBBC021 were 98k. 

![Image](https://github.com/user-attachments/assets/cc88d7f5-986f-4f04-a926-70de235bdc27)

However, on the actual webpage of the dataset, there are only 39,600 images. I'm not quite clear on how you constructed the 98k data. Could you elaborate on this part? 

![Image](https://github.com/user-attachments/assets/c2b4d9b3-3317-4f48-83e9-7ab906bfe203)

Additionally, in BBBC021 dataset, you trained with all the compounds. So, when you do inference, do you still test with the compounds that were trained? This doesn't seem reasonable. Shouldn't you leave out a portion of the compounds for training and then test on this part? Moreover, there is no early stopping mechanism in your training code. In your `log_validation` function, you did not calculate the validation loss. How do you determine whether the model is overfitting?

**Comment by znavidi** [comment]
Please find the response to the raised questions below. For future, please consider opening separate issues for different topics to make it easier for other users to find relevant responses through search.

- In the manuscript, we explained that data augmentation was applied to balance sample sizes across perturbation groups during training, addressing dataset imbalance. Using random flipping and rotation, we increased training images while preventing the model from learning confounding factors like absolute cell orientation and position. Appendix Table 4 (last column) details the target number of augmented images for each perturbation, while the "Training images" column lists the total images used during training.

- Our experiments included training/inference in both in-distribution and out of distribution setting. For in-distribution, cellular morphology in the generated images were validated for perturbations seen during training (such as BBBC021 all compounds). For the BBBC021 including 14 compounds, the training was performed on only 14 perturbation and inference was performed on both seen and unseen compounds, elaborated in different sections of the manuscript.

- We used a fixed number of training steps (with a fixed batch size) for each experiment and conducted regular visual inspections of validation images (using wandb) alongside monitoring the training loss. Out-of-distribution (OOD) prediction is inherently a highly challenging task, and our experiments demonstrated that Morphodiff, as a diffusion-based model, has the potential to capture perturbation-specific cellular morphology in unseen perturbations. Users can adjust the validation function as needed as there is no single stopping criterion.

Hope the above clarifies further. Also, pls pull the updated code if you plan to run scripts from the repository.

**Comment by prsigma** [comment]
Thank you for your answer. I have understood it clearly now!
